### ISSUE TITLE
Mixed precision training fails due to NaN in batch norm running_mean

### ISSUE THREAD

**Issue by caiqi** [pytorch/pytorch#88474]
### 🐛 Describe the bug

During mixed precision training, the NaN gradient is skipped by grad_scaler. However, the NaN output will destroy the batch norm running mean and variance. This is severe because:
1. It can only be observed during validation since during training the running_mean and running_variance are not used.
2. It will break the whole training process of AMP.

Can batch norm skip NaN samples while updating the running_mean and running_var? Is there any way to achieve such a purpose? Following is an MRE. 
```python
import torch 
data = torch.rand(2, 3, 10).cuda()
from torch import nn
from torch.cuda.amp import autocast as autocast

network = nn.Sequential(
    nn.Conv1d(3, 16, 3, padding=1),
    nn.ReLU(),
    nn.BatchNorm1d(16)).cuda()

with autocast():
    data = data.half() * 100000
    output = network[0](data)
    output = network[1](output)
    output = network[2](output)
    print(network[1].running_mean)
```
The output is :
```
tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
       device='cuda:0')
```

### Versions

PyTorch version: 1.10.1
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.31

Python version: 3.9.7 (default, Sep 16 2021, 13:09:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.10.102.1-microsoft-standard-WSL2-x86_64-with-glibc2.31

**Comment by leslie-fang-intel** [comment]
cc @CaoE

**Comment by leslie-fang-intel** [comment]
@caiqi From your example code, I think it may not be a `AMP` or `grad_scaler` specific issue. The `input data` in your example may already exceed the range of `fp16`.

**Comment by caiqi** [comment]
@leslie-fang-intel It is not a problem of AMP or grad_scaler. But fp16 is usually used in AMP training. It would be better if AMP could add the functionality of handling this case: NaNs in output tensor. For example, adding a hook to remove NaN before updating running_mean/running_var?

**Comment by leslie-fang-intel** [comment]
In your example code, the `running_mean/running_var` of `BN` is NaN, because some elements of the input data `data = data.half() * 100000` exceeds the range of `fp16`.
Is it the same case you see in your training use case? If so, could you share more details about the true use case.

**Comment by caiqi** [comment]
@leslie-fang-intel I am training a new 3D object detection model with AMP based on mmdetection3d(https://github.com/open-mmlab/mmdetection3d). During training, I find that both weights and input of some convolution layers are normal but the output exceeds the range of fp16, causing NaN in the batch norm running_mean. The training can still go on since the grad_scaler will skip the NaN gradient and running_mean/running_var are not used during training. But when finished training and evaluating the model, the output is all invalid NaNs.

**Comment by leslie-fang-intel** [comment]
Hi @mcarilli, If my understand is right, the `running_mean/running_var` updating happens in CUDA function or CUDNN like https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Normalization.cu#L369. Will you take a look of it?
